VideoClassification/model.py

```diff
@@ -0,0 +1,267 @@
+      initializer=tf.keras.initializers.TruncatedNormal(
+        stddev=1.0 / math.sqrt(feature_dim)
+      ),
+      trainable=True,
+      name="cluster_centers" + str(num_clusters)
+    )
+    self.feature_dim = feature_dim
+    self.max_frames = input_shape[-2]
+
+  def call(self, frames):
+    """Apply the NetVLAD module to the given frames.
+    Args:
+      frames: A tensor with shape [batch_size, max_frames, feature_dim].
+    Returns:
+      vlad_out: A tensor with shape [batch_size, feature_dim * num_clusters].
+    Raises:
+      ValueError: If the `feature_dim` of input is not defined.
+    """
+    feature_dim = self.feature_dim
+    max_frames = self.max_frames
+    
+    frames = tf.reshape(frames, (-1, feature_dim))
+    activation = self.fc(frames)
+    activation = tf.reshape(activation, (-1, max_frames, self.num_clusters))
+    
+    activation_sum = tf.math.reduce_sum(activation, axis=-2, keepdims=True)
+    cluster_activation = activation_sum * self.cluster_centers
+
+    frames = tf.reshape(frames, (-1, max_frames, feature_dim))
+    activation = tf.transpose(
+      tf.matmul(tf.transpose(activation, perm=(0, 2, 1)), frames), perm=(0, 2, 1)
+    )
+    
+    vlad_out = activation - cluster_activation
+    vlad_out = tf.nn.l2_normalize(vlad_out, 1)
+    vlad_out = tf.reshape(vlad_out, (-1, feature_dim * self.num_clusters))
+    vlad_out = tf.nn.l2_normalize(vlad_out, 1)
+
+    return vlad_out
+
+  def compute_output_shape(self, input_shape):
+    input_shape = tf.TensorShape(input_shape).as_list()
+    return tf.TensorShape([input_shape[0], input_shape[-1] * self.num_clusters])
+
+  def get_config(self):
+    config = {"num_clusters": self.num_clusters}
+    base_config = super().get_config()
+    return dict(list(base_config.items()) + list(config.items()))
+
+class ContextGating(tf.keras.layers.Layer):
+  """Implements the Context Gating Layer from https://arxiv.org/abs/1706.06905
+
+  Input shape:
+  2D tensor with shape: `(batch_size, feature_dim)`.
+  Output shape:
+  2D tensor with shape: `(batch_size, feature_dim)`.
+  """
+
+  def __init__(self, input_shape, **kwargs):
+    super(ContextGating, self).__init__(**kwargs)
+    feature_dim = input_shape[-1]
+    if not isinstance(feature_dim, int):
+      feature_dim = feature_dim.value
+    self.fc = tf.keras.layers.Dense(
+      units=feature_dim,
+      activation=tf.nn.sigmoid,
+      kernel_regularizer=tf.keras.regularizers.l2(1e-5),
+    )
+
+  def call(self, model_input):
+    """Apply the ContextGating module to the given input.
+
+    Args:
+      model_input: A tensor with shape [batch_size, feature_dim].
+    Returns:
+      A tensor with shape [batch_size, feature_dim].
+    Raises:
+      ValueError: If the `feature_dim` of model_input is not defined.
+    """
+    model_input.shape.assert_has_rank(2)
+    feature_dim = model_input.shape.as_list()[-1]
+    if feature_dim is None:
+      raise ValueError("Last dimension must be defined.")
+    
```

Comment:
I find it just helps with spacing out parts of the code that are in different steps. ie. some sections of the code can be grouped together. I find that it can sometimes help with readability, but I have changed it.